### DISCUSSION TITLE
How to measure each test duration (Cypress Cloud)

### DISCUSSION THREAD

**Discussion by szanuje** [cypress-io/cypress#24837] [Questions and Help]
Hey, I recently started using Cypress and it's awesome!
I set up Cypress Cloud and run my first tests. I noticed that in `Run duration` tab I can see the summaries of `cypress run` durations. Is it possible to have metrics of each test durations?

For example I have this test:
```
describe(`test`, () => {
  it("should get API response", () => {
    cy.request({
      method: "POST",
      url: "/api/graphql",
      body: `{
  "query": "some GQL query"
}`,
    }).then((response) => {
      expect(response.status).to.eq(200);
      expect(JSON.stringify(response.body)).to.include("prop");
    });
  });
});
```
It lasts 600ms on my local environment, when run inside Cypress app (`npx cypress open`).
I also would like to extend this test, and wrap into another `describe` function, to execute a test suite, where each test would check different graphQL query.
I am worried if I will be able to get granular metrics - the exact run time of each small test case in order to check my API performance.

**Comment by portyu9** [comment]
Hi @szanuje 

Yes — and for the API-performance part of your question, you actually do not need to derive the duration from the whole Cypress test.

`cy.request()` returns a `duration` value in milliseconds:

```js
cy.request({
  method: 'POST',
  url: '/api/graphql',
  body: {
    query: '...',
  },
}).then((response) => {
  expect(response.status).to.eq(200)

  cy.log(`Request duration: ${response.duration} ms`)

  expect(response.duration).to.be.lessThan(1000)
})
```

That is much more useful for your GraphQL example than measuring the entire `it()` block, because the test duration also includes Cypress commands, assertions, hooks, and other test overhead.

Current Cypress Cloud also has **Project Analytics → Slowest tests**, where results can be viewed by individual test case and include duration/median-duration information. So the Cloud functionality is more granular now than it was when this discussion was opened.

One caveat: `response.duration` measures the request as Cypress sees it. `cy.request()` is made from Cypress's Node process, so this is useful for regression checks but it is not the same thing as pure server execution time or a proper load/performance benchmark.

For strict API performance/SLA testing, server-side timing or a dedicated performance/load-testing tool will give more controlled measurements. For detecting that a GraphQL request suddenly went from ~200 ms to ~2 seconds, `response.duration` is perfectly useful.

Please mark this answer as accepted if it helped, thank you.
